File: mobile/test/features/activity/dm_resurface_test.dart
import 'package:flutter_test/flutter_test.dart';
import 'package:buzz/features/activity/dm_resurface.dart';
import 'package:buzz/shared/relay/relay.dart';

void main() {
  const self =
      '1111111111111111111111111111111111111111111111111111111111111111';
  const alice =
      '2222222222222222222222222222222222222222222222222222222222222222';
  const bob =
      '3333333333333333333333333333333333333333333333333333333333333333';

  test('derives group DM peers from authoritative membership', () {
    expect(dmPeerPubkeysFromMembers([self, alice, bob], self), {alice, bob});
    expect(dmPeerPubkeysFromMembers([alice, bob], self), isEmpty);
  });

  test('accepts external channel messages regardless of p tags', () {
    NostrEvent event({
      int kind = EventKind.streamMessage,
      String? author,
      List<List<String>>? tags,
    }) => NostrEvent(
      id: 'event-1',
      pubkey: author ?? alice,
      createdAt: 1,
      kind: kind,
      tags:
          tags ??
          const [
            ['h', 'dm-1'],
            ['p', self],
          ],
      content: 'hello',
      sig: 'sig',
    );

    expect(isIncomingChannelMessageFromOther(event(), self), isTrue);
    expect(
      isIncomingChannelMessageFromOther(event(kind: EventKind.reaction), self),
      isFalse,
    );
    expect(
      isIncomingChannelMessageFromOther(event(author: self), self),
      isFalse,
    );
    // #h-scoped delivery already guarantees relevance: an untagged DM from
    // another sender still qualifies.
    expect(
      isIncomingChannelMessageFromOther(
        event(
          tags: const [
            ['h', 'dm-1'],
          ],
        ),
        self,
      ),
      isTrue,
    );
  });
}

File: desktop/src/features/channels/dmResurface.test.mjs
import assert from "node:assert/strict";
import test from "node:test";

import {
  dmPeerPubkeysFromMembers,
  isIncomingChannelMessageFromOther,
  markHiddenDmFeedItems,
} from "./dmResurface.ts";

const SELF = "1".repeat(64);
const ALICE = "2".repeat(64);
const BOB = "3".repeat(64);

function item(overrides = {}) {
  return {
    id: "event-1",
    kind: 9,
    pubkey: ALICE,
    content: "hello",
    createdAt: 10,
    channelId: "dm-1",
    channelName: "",
    tags: [
      ["h", "dm-1"],
      ["p", SELF],
      ["p", BOB],
    ],
    category: "mention",
    ...overrides,
  };
}

function relayEvent(overrides = {}) {
  return {
    id: "event-1",
    kind: 40002,
    pubkey: ALICE,
    content: "hello",
    created_at: 10,
    tags: [
      ["h", "dm-1"],
      ["p", SELF],
      ["p", BOB],
    ],
    sig: "",
    ...overrides,
  };
}

test("DM resurface derives peers from authoritative membership", () => {
  const members = [{ pubkey: SELF }, { pubkey: ALICE }, { pubkey: BOB }];
  assert.deepEqual(dmPeerPubkeysFromMembers(members, SELF), [ALICE, BOB]);
  assert.deepEqual(dmPeerPubkeysFromMembers([{ pubkey: ALICE }], SELF), []);
});

test("only external channel messages qualify, regardless of p tags", () => {
  // #h-scoped delivery already guarantees relevance, so eligibility no longer
  // requires a self `p` tag — an untagged DM from another sender still counts.
  assert.equal(isIncomingChannelMessageFromOther(relayEvent(), SELF), true);
  assert.equal(
    isIncomingChannelMessageFromOther(relayEvent({ kind: 7 }), SELF),
    false,
  );
  assert.equal(
    isIncomingChannelMessageFromOther(relayEvent({ pubkey: SELF }), SELF),
    false,
  );
  assert.equal(
    isIncomingChannelMessageFromOther(relayEvent({ tags: [] }), SELF),
    false,
  );
  assert.equal(
    isIncomingChannelMessageFromOther(
      relayEvent({ tags: [["h", "dm-1"]] }),
      SELF,
    ),
    true,
  );
});

test("hidden feed items are projected as DMs for Inbox presentation", () => {
  const feed = {
    feed: {
      mentions: [item()],
      needsAction: [],
      activity: [],
      agentActivity: [],
    },
    meta: { since: 0, total: 1, generatedAt: 10 },
  };
  const marked = markHiddenDmFeedItems(feed, new Set(["dm-1"]));
  assert.equal(marked.feed.mentions[0].channelType, "dm");
});
